aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -327,27 +332,72 @@ SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const S
         }
     );
 
+    if (r.scalar_type() != commonDtype) {
+      r_values = r_values.to(r.scalar_type());
+    }
+    get_sparse_impl(r)->set_indices_and_values_unsafe(r_indices, r_values);
     get_sparse_impl(r)->set_nnz_and_narrow(r_i);
+
     // TODO: I think it may be possible to track inside the loop and
     // detect when we are uncoalesced (e.g., by observing that an
     // index goes backwards) which may be more precise than using the
     // coalesced flag here.  But this is easy.
-    return r._coalesced_(t_coalesced && s_coalesced);
-  } else {
+    return r._coalesced_(coalesced);
+}
+
+SparseTensor& add_out_sparse_non_contiguous(SparseTensor& r, const SparseTensor& t, const SparseTensor& src, Scalar value, ScalarType commonDtype) {
+    Tensor t_values = t._values().to(commonDtype);
+    Tensor s_values = src._values().to(commonDtype);
+
     // If `t` or `src` contains non-contiguous `values`, `THBlas_axpy` doesn't work
     // and we concat the indices and values tensors instead.
     AT_DISPATCH_ALL_TYPES(
-      s_values.scalar_type(), "add_out_sparse_cuda", [&] {
+      commonDtype, "add_out_sparse_cuda", [&] {
           if (value.to<scalar_t>() != static_cast<scalar_t>(1)) {
             s_values = s_values.mul(value);
           }
         });
 
-    LongTensor r_indices = at::cat({t_indices, src_indices}, 1);
-    Tensor r_values = at::cat({t_values, s_values}, 0);
+    LongTensor r_indices = at::cat({t._indices(), src._indices()}, 1);
+    Tensor r_values = at::cat({t_values, s_values}, 0).to(r.scalar_type());
     alias_into_sparse(r, r_indices, r_values);
 
     return r;
+}
+
+Tensor& add_out_dense_sparse_cpu(Tensor& r, const Tensor& dense, const SparseTensor& sparse_, Scalar value);
+
+SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const SparseTensor& src, Scalar value) {
+  if (!t.is_sparse()) {
```

Comment:
Why is `t` a SparseTensor if it's not sparse?

Reply:
That's existing code, only moved not changed. It works because `using SparseTensor = Tensor;`

Reply:
Do you know why we do that? It might be nicer to change the declaration to use `Tensor` instead, if `SparseTensor` is used just for style purposes.

Reply:
I have no idea, but if changing it I'd prefer to do that in a separate PR, since unrelated.

Reply:
Since we're modifying this anyways, it's a quick fix especially if it is correct. But feel free to file an issue and follow up on it.

Reply:
> It might be nicer to change the declaration to use Tensor instead, if SparseTensor is used just for style purposes.

I do assume it's for style/annotation purposes. I think changing it to `Tensor` could be incorrect/confusing though. This function only handles add(sparse, sparse), and if you give it a `(dense, sparse)`, it redirects to a function that has the `(dense, sparse)` signature. It's not necessarily a stylistic choice I would have made, but I also don't think it's wrong as-is.

Reply:
Right, that's all I'm saying is it's an orthogonal style choice across the whole file. I don't think we should change this one location -- if we're going to get rid of it we should remove all uses. I know it's confusing to see the ability to pass a dense tensor to a function annotated to handle sparse, but if we simply change it to `Tensor` in this one place then the annotations become inconsistent across the file and someone might think this function handles dense.

It may also be possible to get rid of these conditions/checks (or replace torch checks with internal asserts) since there have been more recent changes to dispatch (in native_functions we now have lines like `SparseCPU: add_sparse_` so this shouldn't be called incorrectly), but it'd take a little investigation to decide if this annotation is desirable and supportable in a more readable way, so I think we should consider it separately.